cli/save.go

```diff
@@ -0,0 +1,79 @@
+package main
+
+import (
+	"context"
+	"io"
+	"os"
+
+	"github.com/spf13/cobra"
+)
+
+// saveDescription is used to describe save command in detail and auto generate command doc.
+var saveDescription = "save an image to a tar archive."
+
+// SaveCommand use to implement 'save' command.
+type SaveCommand struct {
+	baseCommand
+	output string
+}
+
+// Init initialize save command.
+func (save *SaveCommand) Init(c *Cli) {
+	save.cli = c
+	save.cmd = &cobra.Command{
+		Use:   "save [OPTIONS] IMAGE [IMAGE...]",
+		Short: "Save an image to a tar archive or STDOUT",
+		Long:  saveDescription,
+		Args:  cobra.ExactArgs(1),
+		RunE: func(_ *cobra.Command, args []string) error {
+			return save.runSave(args)
+		},
+		Example: saveExample(),
+	}
+	save.addFlags()
+}
+
+// addFlags adds flags for specific command.
+func (save *SaveCommand) addFlags() {
+	flagSet := save.cmd.Flags()
+	flagSet.StringVarP(&save.output, "output", "o", "", "Save to a tar archive file, instead of STDOUT")
+}
+
+// runSave is the entry of save command.
+func (save *SaveCommand) runSave(args []string) error {
+	ctx := context.Background()
+	apiClient := save.cli.Client()
+
+	r, err := apiClient.ImageSave(ctx, args[0])
+	defer r.Close()
+	if err != nil {
+		return err
+	}
+
+	out := os.Stdout
+	if save.output != "" {
+		out, err = os.Create(save.output)
+		if err != nil {
+			return nil
+		}
+	}
+
+	if _, err := io.Copy(out, r); err != nil {
+		return err
+	}
+	if err := out.Close(); err != nil {
```

Comment:
if out is os.Stdout, we should not close it

Reply:
Emmm, I'm just curious to ask why we shouldn't close client stdout...

Reply:
@shaloulcy I'm just curious to ask why we shouldn't close client stdout...

Reply:
if we did like

```
func main() {
        out := os.Stdout
        fmt.Println(out.Close())
        fmt.Println("haha")
}
```
The following stdout will missing so that we can't close the stdout.